### PR TITLE
Use callbacks for Julia v0.6

### PR DESCRIPTION
**Pull request by vchuravy**
JuliaLang/julia#21441 adds the ability to register callbacks to `Base.require` removing the need to overwrite the function. 

We can probably simplify Requires a bit more and add more tests.

### CODE REVIEW

**Comment by MikeInnes** [comment]
That's awesome. This will get rid of the ugly warning on startup as well.

Is this good to merge as far as you're concerned? Would appreciate some testing from @davidanthoff to see if this resolves his issues.

File: src/require.jl [hunk 1]
```diff
@@ -1,35 +1,23 @@
...
-# Since `Base.require` uses `eval` it switches automatically into
-# a newer world and call this `Base.require` from there.
-# We need to repeatedly switch back into the old world.
-@init @guard begin
-  function Base.require(mod::Symbol)
-    Main.Requires.oldcall(Base.require, mod)
-    Main.Requires.loadmod(string(mod))
-  end
+@init begin
+  push!(Base.package_callbacks, loadmod)
```
Comment (tkelman) [inline src/require.jl]: this needs to set as a minimum version the merge commit for this, or it won't load

**Comment by davidanthoff** [comment]
Yes, that seems to fix the issues I have seen. Hurray!

File: src/require.jl (hunk 1 above)
Comment (vchuravy) [inline src/require.jl]: You mean in the `REQUIRE` file?

Comment (tkelman) [inline src/require.jl]: yes

Comment (davidanthoff) [inline src/require.jl]: Why? There are lots and lots of packages out there that have ``julia 0.6-`` in ``REQUIRE``, and I'm sure many of them rely on features that were introduced somewhere midway through the 0.6 cycle. How is this different here? Once julia 0.6 is released the version bound will be just ``0.6`` I assume in any case, right?

**Comment by vchuravy** [comment]
I think this is good to go from my side and we can work on improving tests later.

File: src/require.jl (hunk 1 above)
Comment (vchuravy) [inline src/require.jl]: Because we don't want to break the Requires.jl installation off people on v0.6-beta (see travis).

Comment (tkelman) [inline src/require.jl]: There are only 25 registered packages that have any tagged versions that require `julia 0.6-`, and I'd wager that nearly all of those were tagged significantly more than a day after the feature they rely on made it into Julia master. When there are earlier versions of the package available that don't rely on the feature, being specific about REQUIRE dependencies allows Pkg to do the right thing and use earlier versions when those are more likely to work.

File: REQUIRE [hunk 2]
```diff
@@ -1 +1 @@
-julia 0.6-
+julia 0.6.0-pre.beta.450
```
Comment (vchuravy) [inline REQUIRE]: @tkelman Is this format correct? I took the merge commit and used `contrib/commit-name.sh`

Comment (tkelman) [inline REQUIRE]: yep

**Comment by davidanthoff** [comment]
Great! Can you tag a version?

**Pull request merged**